lib/internal/modules/esm/module_job.js

```diff
@@ -92,7 +97,39 @@ class ModuleJob {
         this.module.instantiate();
       }
     } catch (e) {
-      decorateErrorStack(e);
+      if (StringPrototypeIncludes(e.message,
+                                  ' does not provide an export named')) {
+        const start = StringPrototypeIndexOf(e.message, '\'') + 1;
+        const end =
+          StringPrototypeIndexOf(e.message,
+                                 ' does not provide an export named') - 1;
+        const childSpecifier = StringPrototypeSlice(e.message, start, end);
+        let childFileURL;
+        if (StringPrototypeStartsWith(childSpecifier, '.') ||
+          StringPrototypeStartsWith(childSpecifier, '/')) {
+          // If the specifier is a relative or absolute path make a new
+          // URL object to resolve the file url. This will not work for
+          // bare specifiers.
+          childFileURL = (new URL(childSpecifier, this.module.url)).href;
+        } else {
+          // If the specifier is a bare specifier use `this.loader.resolve`
+          // to resolve the file url. Using this.loader.resolve with
+          // relative or absolute path will resolve the file url with
+          // a base of the root node, not the parent.
+          // TODO (myles) ensure that this works as expected for deeply
+          // imported modules. This edge case is highly unlikely, but
+          // probably best to test for it.
+          childFileURL =
+            await this.loader.resolve(childSpecifier, this.module.url);
```

Comment:
Same as above... if there are nested node_modules folders and duplicated entries it might be possible to resolve the wrong bare import.

The reason we can't use `this.loader.resolve` for the above case is that it checks for the existence of the file when resolving. So in theory once we get the correct base url here we should be able to remove all the comments and the if / else statement and use this.loader.resolve for any child specifier.